File: plugins/gradle/tooling-extension-v1.9/src/org/jetbrains/plugins/gradle/tooling/v1_9/builder/WarModelBuilderImpl.java
/*
 * Copyright 2000-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jetbrains.plugins.gradle.tooling.v1_9.builder;

import org.gradle.api.Action;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.file.FileVisitDetails;
import org.gradle.api.file.FileVisitor;
import org.gradle.api.internal.file.copy.CopySpecInternal;
import org.gradle.api.java.archives.Manifest;
import org.gradle.api.plugins.WarPlugin;
import org.gradle.api.tasks.bundling.War;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.plugins.gradle.model.WebConfiguration;
import org.jetbrains.plugins.gradle.tooling.ModelBuilderService;
import org.jetbrains.plugins.gradle.tooling.internal.WarModelImpl;
import org.jetbrains.plugins.gradle.tooling.internal.WebConfigurationImpl;
import org.jetbrains.plugins.gradle.tooling.internal.WebResourceImpl;

import java.io.File;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Vladislav.Soroka
 * @since 11/5/13
 */
public class WarModelBuilderImpl implements ModelBuilderService {

  private static final String WEB_APP_DIR_PROPERTY = "webAppDir";
  private static final String WEB_APP_DIR_NAME_PROPERTY = "webAppDirName";

  @Override
  public boolean canBuild(String modelName) {
    return WebConfiguration.class.getName().equals(modelName);
  }

  @Nullable
  @Override
  public Object buildAll(String modelName, Project project) {
    final WarPlugin warPlugin = project.getPlugins().findPlugin(WarPlugin.class);
    if (warPlugin == null) return null;

    final String webAppDirName = !project.hasProperty(WEB_APP_DIR_NAME_PROPERTY) ?
                                 "src/main/webapp" : String.valueOf(project.property(WEB_APP_DIR_NAME_PROPERTY));

    final File webAppDir = !project.hasProperty(WEB_APP_DIR_PROPERTY)
                           ? new File(project.getProjectDir(), webAppDirName)
                           : (File)project.property(WEB_APP_DIR_PROPERTY);


    List<WebConfiguration.WarModel> warModels = new ArrayList<WebConfiguration.WarModel>();


    for (Task task : project.getTasks()) {
      if (task instanceof War) {
        final WarModelImpl warModel =
          new WarModelImpl(((War)task).getArchiveName(), webAppDirName, webAppDir);
        final List<WebConfiguration.WebResource> webResources = new ArrayList<WebConfiguration.WebResource>();

        final War warTask = (War)task;
        warModel.setWebXml(warTask.getWebXml());

        warTask.getRootSpec().walk(new Action<CopySpecInternal>() {
          @Override
          public void execute(CopySpecInternal internal) {
            final String relativePath = internal.getDestPath().getPathString();
            internal.getSource().visit(new FileVisitor() {
              @Override
              public void visitDir(FileVisitDetails dirDetails) {
                try {
                  addPath(webResources, relativePath, dirDetails.getPath(), dirDetails.getFile());
                }
                catch (Exception ignore) {
                }
              }

              @Override
              public void visitFile(FileVisitDetails fileDetails) {
                try {
                  if (warTask.getWebXml() == null ||
                      !fileDetails.getFile().getCanonicalPath().equals(warTask.getWebXml().getCanonicalPath())) {
                    addPath(webResources, relativePath, fileDetails.getPath(), fileDetails.getFile());
                  }
                }
                catch (Exception ignore) {
                }
              }
            });
          }
        });

        warModel.setWebResources(webResources);
        warModel.setClasspath(warTask.getClasspath().getFiles());

        Manifest manifest = warTask.getManifest();
        if (manifest != null) {
          StringWriter writer = new StringWriter();
          manifest.writeTo(writer);
          warModel.setManifestContent(writer.toString());
        }
        warModels.add(warModel);
      }
    }

    return new WebConfigurationImpl(warModels);
  }

  private static void addPath(List<WebConfiguration.WebResource> webResources, String warRelativePath, String fileRelativePath, File file) {
    warRelativePath = warRelativePath == null ? "" : warRelativePath;

    WebConfiguration.WebResource webResource = new WebResourceImpl(warRelativePath, fileRelativePath, file);
    webResources.add(webResource);
  }
}

File: plugins/gradle/tooling-extension-impl/src/org/jetbrains/plugins/gradle/tooling/builder/WarModelBuilderImpl.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.plugins.gradle.tooling.builder

import com.intellij.gradle.toolingExtension.impl.modelBuilder.Messages
import com.intellij.gradle.toolingExtension.impl.util.GradleTaskUtil
import com.intellij.gradle.toolingExtension.util.GradleVersionUtil
import org.gradle.api.Project
import org.gradle.api.file.FileVisitDetails
import org.gradle.api.java.archives.internal.ManifestInternal
import org.gradle.api.plugins.WarPlugin
import org.gradle.api.tasks.bundling.War
import org.jetbrains.plugins.gradle.model.web.WebConfiguration
import org.jetbrains.plugins.gradle.model.web.WebConfiguration.WebResource
import org.jetbrains.plugins.gradle.tooling.AbstractModelBuilderService
import org.jetbrains.plugins.gradle.tooling.Message
import org.jetbrains.plugins.gradle.tooling.ModelBuilderContext
import org.jetbrains.plugins.gradle.tooling.internal.web.WarModelImpl
import org.jetbrains.plugins.gradle.tooling.internal.web.WebConfigurationImpl
import org.jetbrains.plugins.gradle.tooling.internal.web.WebResourceImpl
import java.io.ByteArrayOutputStream
import java.io.File

private val WAR_MODEL_NAME = WebConfiguration::class.java.name

private const val WEB_APP_DIR_PROPERTY = "webAppDir"
private const val WEB_APP_DIR_NAME_PROPERTY = "webAppDirName"

private val is82OrBetter = GradleVersionUtil.isCurrentGradleAtLeast("8.2")

class WarModelBuilderImpl : AbstractModelBuilderService() {

  override fun canBuild(modelName: String): Boolean {
    return WAR_MODEL_NAME == modelName
  }

  override fun buildAll(modelName: String, project: Project, context: ModelBuilderContext): Any? {
    project.plugins.findPlugin(WarPlugin::class.java) ?: return null

    val warModels = mutableListOf<WarModelImpl>()

    for (task in project.tasks.withType(War::class.java)) {

      val archiveFile = GradleTaskUtil.getTaskArchiveFile(task)
      val archiveName = GradleTaskUtil.getTaskArchiveFileName(task)

      val (webAppDir, webAppDirName) = collectWebAppDirAndName(project, task)
      val webXml = task.webXml
      val webResources = collectWebResources(context, project, task, webXml)
      val classpath = collectClasspath(context, project, task)
      val manifestContent = collectManifestContent(task)

      val warModel = WarModelImpl(archiveName, webAppDirName, webAppDir)
      warModel.archivePath = archiveFile
      warModel.webXml = webXml
      warModel.webResources = webResources
      warModel.classpath = classpath
      warModel.manifestContent = manifestContent
      warModels.add(warModel)
    }

    return WebConfigurationImpl(warModels)
  }

  private fun reportErrorMessage(context: ModelBuilderContext, project: Project, exception: Exception) {
    reportErrorMessage(WAR_MODEL_NAME, project, context, exception)
  }

  override fun reportErrorMessage(modelName: String, project: Project, context: ModelBuilderContext, exception: Exception) {
    context.messageReporter.createMessage()
      .withGroup(Messages.WAR_CONFIGURATION_MODEL_GROUP)
      .withKind(Message.Kind.WARNING)
      .withTitle("JEE project import failure")
      .withText("Web Facets/Artifacts may not be configured properly")
      .withException(exception)
      .reportMessage(project)
  }

  private fun collectWebAppDirAndName(project: Project, task: War): Pair<File, String> {
    if (is82OrBetter) {
      val webAppDir = task.webAppDirectory.asFile.get()
      val webAppDirName = webAppDir.name
      return webAppDir to webAppDirName
    }

    val webAppDirName = project.findProperty(WEB_APP_DIR_NAME_PROPERTY)?.toString()
                        ?: "src/main/webapp"
    val webAppDir = project.findProperty(WEB_APP_DIR_PROPERTY) as? File
                    ?: File(project.projectDir, webAppDirName)
    return webAppDir to webAppDirName
  }

  private fun collectWebResources(
    context: ModelBuilderContext,
    project: Project,
    task: War,
    webXml: File?,
  ): List<WebResource> {
    try {
      val webResources = ArrayList<WebResource>()
      CopySpecWalker.walk(task.rootSpec, object : CopySpecWalker.Visitor {
        override fun visitSourcePath(relativePath: String?, path: String) {
          val file = File(path).takeIf { it.isAbsolute } ?: File(task.project.projectDir, path)
          webResources.addWebResource(relativePath, "", file)
        }

        override fun visitDir(relativePath: String?, dirDetails: FileVisitDetails) {
          webResources.addWebResource(relativePath, dirDetails.path, dirDetails.file)
        }

        override fun visitFile(relativePath: String?, fileDetails: FileVisitDetails) {
          val file = fileDetails.file
          if (webXml == null || file.canonicalPath != webXml.canonicalPath) {
            webResources.addWebResource(relativePath, fileDetails.path, file)
          }
        }
      })
      return webResources
    }
    catch (e: Exception) {
      reportErrorMessage(context, project, e)
    }
    return emptyList()
  }

  private fun MutableList<WebResource>.addWebResource(
    warDirectory: String?,
    relativePath: String,
    file: File,
  ) {
    add(WebResourceImpl(warDirectory ?: "", relativePath, file))
  }

  private fun collectClasspath(context: ModelBuilderContext, project: Project, task: War): Set<File> {
    try {
      val classpath = task.classpath ?: return emptySet()
      return classpath.files.toSet()
    }
    catch (e: Exception) {
      reportErrorMessage(context, project, e)
    }
    return emptySet()
  }

  private fun collectManifestContent(task: War): String? {
    val manifest = task.manifest
    if (manifest !is ManifestInternal) return null
    val outputStream = ByteArrayOutputStream()
    manifest.writeTo(outputStream)
    val contentCharset = manifest.contentCharset
    return outputStream.toString(contentCharset)
  }
}